data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/client/CloudWatchLogsService.java

```diff
@@ -23,158 +19,134 @@
-                int logLength = logJsonString.length();
+    private void stageLogEvents() {
+        sinkStopWatch.stopAndResetStopWatch();
 
-                if (thresholdCheck.isGreaterThanMaxEventSize(logLength + APPROXIMATE_LOG_EVENT_OVERHEAD_SIZE)) {
-                    LOG.warn("Event blocked due to Max Size restriction! {Event Size: " + (logLength + APPROXIMATE_LOG_EVENT_OVERHEAD_SIZE) + " bytes}");
-                    continue;
-                }
+        ArrayList<byte[]> eventMessageClone = new ArrayList<>();
+        cloneLists(buffer.getBufferedData(), eventMessageClone);
 
-                int bufferSizeWithOverhead = (buffer.getBufferSize() + (buffer.getEventCount() * APPROXIMATE_LOG_EVENT_OVERHEAD_SIZE));
-                if ((thresholdCheck.isGreaterThanThresholdReached(sinkStopWatch.getStopWatchTimeSeconds(),  bufferSizeWithOverhead + logLength + APPROXIMATE_LOG_EVENT_OVERHEAD_SIZE, buffer.getEventCount() + 1) && (buffer.getEventCount() > 0))) {
-                    pushLogs();
-                }
+        ThreadTaskEvents dataToPush = new ThreadTaskEvents(eventMessageClone, bufferedEventHandles);
+        taskQueue.add(dataToPush);
 
-                if (log.getData().getEventHandle() != null) {
-                    bufferedEventHandles.add(log.getData().getEventHandle());
-                }
-                buffer.writeEvent(logJsonString.getBytes());
-            }
+        buffer.clearBuffer();
+        bufferedEventHandles = new ArrayList<>();
 
-            runExitCheck();
+        CloudWatchLogsDispatcher newTaskDispatcher = CloudWatchLogsDispatcher.builder()
+                .taskQueue(taskQueue)
+                .cloudWatchLogsClient(cloudWatchLogsClient)
+                .cloudWatchLogsMetrics(cloudWatchLogsMetrics)
+                .logGroup(logGroup)
+                .logStream(logStream)
+                .retryCount(retryCount)
+                .backOffTimeBase(backOffTimeBase)
+                .build();
 
-            return true;
-        } catch (InterruptedException e) {
-            LOG.error("Caught InterruptedException while attempting to publish logs!");
-            return false;
-        }
+        asyncExecutor.execute(newTaskDispatcher);
     }
 
-    private void pushLogs() throws InterruptedException {
-        sinkStopWatch.stopAndResetStopWatch();
-        sinkStopWatch.startStopWatch();
-
-        boolean succeededTransmission = logPusher.pushLogs(buffer, cloudWatchLogsClient, logGroup, logStream);
-        releaseEventHandles(succeededTransmission);
-
-        if (!succeededTransmission) {
-            throw new RetransmissionLimitException("Error, timed out trying to push logs! (Max retry_count reached: {" + retryCount + "})");
+    private void addToBuffer(final Record<Event> log) {
+        if (log.getData().getEventHandle() != null) {
+            bufferedEventHandles.add(log.getData().getEventHandle());
         }
+        buffer.writeEvent(log.getData().toJsonString().getBytes());
```

Comment:
We are calling `.toJsonString()` twice. It is a somewhat slow method as it has to convert the object to a JSON string.

You can call `.toJsonString()` once at the top. Then pass in the string in this method so only one call is needed.